### DISCUSSION TITLE
darts interface vs sktime interface

### DISCUSSION THREAD

**Discussion by fkiraly** [alan-turing-institute/sktime#3842] [General]
Answering a question from discord here

> I am looking to migrate my codebase from using darts to sktime. One thing I'm strugging to find documentation on is for a scenario where I have Multiple TIme-Series (each with multiple covariate to predict each time series). Is there something that can do this? I see others have asked similar questions a number of months ago but I couldn't find a Sktime equivalent solution to this: https://unit8co.github.io/darts/examples/01-multi-time-series-and-covariates.html#

Hope I can help with this by putting the interface points on top of each other.

Variable naming and interface:

* darts "target time series" are called "endogen(e)ous variables" in `sktime`, and correspond to the argument `y` in `fit`, `update`, etc., in line with `statsmodels` or the R `forecast` package.
* darts "covariate time series" are called "exogene(e)ous variables" in `sktime`, and correspond to the argument `X` in `fit`, `predict`, `update`
* `sktime` does not distinguish past and future covariates on the interface level, this is a conscious decision to keep the design simple - "past covariates" are simply those that do not have observations in `fit`, of the future. Individual models that make this distinction have the selection of past or future covariates as hyperparameters (in the constructor).
* An example of an estimator for which the past/future distinction is relevant is the `ForecastX` pipeline element, with it you can forecast what darts calls "past covariates" and treat all covariates as "future covariates" by another forecaster.

Multiple time series and multivariate forecasting:

* all `sktime` forecasters support forecasting on multiple time series (panel data, hierarchical data) and multivariate forecasting. If the forecaster does not support this natively, it is applied to multiple variables or instances by a simple loop over those.
* multiple variables and multiple instances are encoded in the `sktime` machine type formalism which allows use of many different data container formats, e.g., `pd.DataFrame` or `numpy` 2D and 3D arrays. In the `pandas` based machine types, instances are multi-index levels over the rows. Variables are columns.
* For an overview of hierarchical forecasting (multiple instances), see [this the global/hierarchical forecasting tutorial](https://github.com/sktime/sktime/blob/main/examples/01c_forecasting_hierarchical_global.ipynb)
* for an overview of multivariate forecasting see the [respective section in the basic forecasting tutorial](https://github.com/sktime/sktime/blob/main/examples/01_forecasting.ipynb)
* for an overview of data formats that `sktime` understands, and how it treats multivariate and multiple instances (panel, hierarchical) in general, also see the [documentation on data types](https://github.com/sktime/sktime/blob/main/examples/AA_datatypes_and_datasets.ipynb)

The tutorials are also available on binder (follow the link in the binder badge in the readme) or in the sktime [user guides](https://www.sktime.org/en/stable/user_guide.html#user-guide). Some of these have [video recordings](https://www.sktime.org/en/stable/tutorials.html#tutorials).

**Comment by fkiraly** [comment]
A note on implemented models:

* `sktime` has a large libary of classical statistical algorithms, ML based algorithms, data preprocessing steps, and composition steps
* `darts` has focused more on "popular" neural network algorithms as far as I can tell. `sktime` does not have these interfaced currently, but this is not an interface limitation of `sktime` (see the discussion above), only no one has interfaced these yet.
* there was an issue on building a general `darts` adapter here, I think that went dormant but there is some very useful code by @ngupta23: https://github.com/sktime/sktime/issues/1624
* it was a while ago so it can have changed, but I think the verdict was that `darts` seemed difficult to interface because the interface was not very unified across the different algorithms, so reimplementation in the `sktime` interface may be an alternative to consider
* there is currently a workstream ongoing, driven by @achieveordie and @AurumnPegasus, in which we want to implement, migrate, or interface forecasting DL algorithms. We have already successfully migrated most of the old `sktime-dl` package into `sktime`, but these were time series classification deep learning models (not forecasting)

Either way, contributions to fill the gap in forecasting DL models would be very much appreciated!
(anyone interested to work on this is welcome to say hello on the slack channel `#contributors`!
familiarity with DL packages is advised)

**Comment by HarveyUdon** [comment]
@fkiraly update: I removed the swear words from your message
---

Who is this and what do you mean?

**Reply by fkiraly** [reply]
Just saw this, @HarveyUdon. I'm confused, what is this? Seems like a notification accident.
